proposals/p1885.md

```diff
@@ -0,0 +1,710 @@
+value. See [Future work][#future-work] for other examples.
+
+#### R-value containers
+
+R-values are likely to be limited in terms of addressing and mutation. The
+impact is undefined until the
+[corresponding design](https://github.com/carbon-language/carbon-lang/pull/2006/files)
+is finalized.
+
+The cursor approach used for `Iterate` has the slight advantage of not requiring
+pointers (and addressing) for some container types. This may be beneficial as a
+first iteration, until the dependent design is updated, at which point we will
+have to define how to handle this special case.
+
+#### Synthetic data
+
+`Next()` can be trivially implemented to return a single computed value
+on-demand based on the cursor, without the need for storage. If the sequence is
+infinite, `Next()` will never return an empty `Optional`, and the `for` loop
+will continue until interrupted by the user.
+
+```carbon
+class SquaredSequence {
+  impl as Iterate where .ElementType = i32 and .CursorType = i32 {
+    fn NewCursor[self: Self]() -> i32 { return -1; }
+    fn Next[self: Self](cursor: i32*) -> Optional(i32) {
+      // Advance and return computed value
+      *cursor = *cursor + 1;
+      return Optional(i32).Create((*cursor) * (*cursor));
+    }
+  }
+}
+```
+
+#### Large containers
+
+The proposal does not present any limitation regarding large collections. The
+cursor can be adapted to the collection size, and no copy of the container
+should occur both for immutable collections, and future mutable elements.
+
+#### Filtering, transforming, and reversing
+
+Provided the traversal order is the same, generic filters and transformers are
+supported by implementing a proxy `Iterate` interface, internally or externally.
+
+This example illustrates a simple proxy interface.
+
+```carbon
+class BasicFilter(T:! Iterate) {
+  var seq: T*;
+  var value: T.ElementType;
+  impl as Iterate where .ElementType = T.ElementType and .CursorType = T.CursorType {
+    fn NewCursor[self: Self]() -> T.ElementType { return self.seq->NewCursor(); }
+    fn Next[self: Self](cursor: T.CursorType*) -> Optional(T.ElementType) {
+      var res: auto = self.seq->Next(cursor);
+      while (res.has_value() && res.get() == value) {
+        res = self.seq->Next(cursor);
+      }
+      return res;
+    }
+  }
+}
+
+fn Main() -> i32 {
+  let container: MyIntContainer = MyIntContainer.Create(0,1,0,2,3,0);
+  let no_zeros: BasicFilter(MyIntContainer) =
+      { .seq = &container, .value = 0 };
+  // Prints "123"
+  for (v: i32 in no_zeros) {
+    Print("{0}", v);
+  }
+  return 0;
+}
+```
+
+On the other hand, changing how the collection is traversed (for example
+reverse), requires internal knowledge about the data layout, and a custom
+implementation. One approach for containers that support reverse iteration would
+be to implement a standard `IterateBidirectional` interface with a `Prev()`
+method. A proxy could be exposed to make it usable with `for` directly.
```

Comment:
Is the `IterateBidirectional` interface part of the proposal or future work?

Reply:
Clarified this is an **example**